File: src/main/java/de/hanno/hpengine/shader/PersistentMappedBuffer.java
package de.hanno.hpengine.shader;

import org.lwjgl.BufferUtils;
import org.lwjgl.opengl.GL43;

import java.nio.DoubleBuffer;
import java.nio.FloatBuffer;

import static org.lwjgl.opengl.GL30.glMapBufferRange;

public class PersistentMappedBuffer extends AbstractPersistentMappedBuffer<DoubleBuffer> {

    public PersistentMappedBuffer(int capacityInBytes) {
        this(capacityInBytes, GL43.GL_SHADER_STORAGE_BUFFER);
    }

    public PersistentMappedBuffer(int capacityInBytes, int target) {
        super(target);
        setCapacityInBytes(capacityInBytes);
    }

    @Override
    protected DoubleBuffer mapBuffer(int capacityInBytes, int flags) {
        DoubleBuffer newBuffer = glMapBufferRange(target, 0, capacityInBytes, flags, BufferUtils.createByteBuffer(capacityInBytes * getPrimitiveSizeInBytes())).asDoubleBuffer();
        if(buffer != null) {
            newBuffer.put(buffer);
            newBuffer.rewind();
        }
        return newBuffer;
    }

    @Override
    public FloatBuffer getValuesAsFloats() {
        FloatBuffer result = BufferUtils.createFloatBuffer(buffer.capacity());
        for(int i = 0; i < buffer.capacity(); i++) {
            result.put(i, (float) buffer.get(i));
        }

        result.rewind();
        return result;
    }

    @Override
    public int getPrimitiveSizeInBytes() {
        return 8;
    }

    @Override
    public DoubleBuffer getValues() {
        return getValues(0, buffer.capacity());
    }

    @Override
    public DoubleBuffer getValues(int offset, int length) {
        DoubleBuffer result = BufferUtils.createDoubleBuffer(length);
        for(int i = 0; i < length; i++) {
            result.put(i, buffer.get(offset+i));
        }

        result.rewind();
        return result;
    }

    @Override
    public void putValues(FloatBuffer values) {
        putValues(0, values);
    }

    @Override
    public void putValues(DoubleBuffer values) {
        putValues(0, values);
    }

    @Override
    public void putValues(int offset, FloatBuffer values) {
        if(values.capacity() > getSizeInBytes()) { setSizeInBytes(values.capacity());}
        bind();
        values.rewind();
        for(int i = offset; i < values.capacity(); i++) {
            buffer.put(i, values.get(i));
        }
        buffer.reset();
    }

    @Override
    public void putValues(int offset, DoubleBuffer values) {
        if(values == buffer) { return; }
        if(values.capacity() > getSizeInBytes()) { setSizeInBytes((offset + values.capacity() )* getPrimitiveSizeInBytes());}
        bind();
        values.rewind();
        buffer.position(offset);
        buffer.put(values);
    }

    @Override
    public void putValues(float... values) {
        putValues(0, values);
    }

    @Override
    public void putValues(int offset, float... values) {
        bind();
        buffer.position(offset);
        double[] doubleValues = new double[values.length];
        for(int i = 0; i < values.length; i++) {
            doubleValues[i] = values[i];
        }
        buffer.put(doubleValues);
    }

    @Override
    public void putValues(int offset, double... values) {
        bind();
        buffer.position(offset);
        buffer.put(values);
    }

    @Override
    public void put(Bufferable... bufferable) {
        put(0, bufferable);
    }

    @Override
    public void put(int offset, Bufferable... bufferable) {
        if(bufferable.length == 0) { return; }
        setCapacityInBytes(bufferable[0].getElementsPerObject() * getPrimitiveSizeInBytes() * bufferable.length);

        buffer.rewind();
        int currentOffset = 0;
        for (int i = 0; i < bufferable.length; i++) {
            Bufferable currentBufferable = bufferable[i];
            setCapacityInBytes((offset+currentOffset + currentBufferable.getElementsPerObject()) * getPrimitiveSizeInBytes());
            double[] currentBufferableArray = currentBufferable.get();
            for (int z = 0; z < currentBufferableArray.length; z++) {
                buffer.put(offset+currentOffset + z, currentBufferableArray[z]);
            }
            currentOffset += currentBufferable.getElementsPerObject();
        }
    }

}

File: opengl/src/main/kotlin/de/hanno/hpengine/graphics/buffer/PersistentMappedBuffer.kt
package de.hanno.hpengine.graphics.buffer

import de.hanno.hpengine.SizeInBytes
import de.hanno.hpengine.buffers.copyTo
import de.hanno.hpengine.graphics.GraphicsApi
import de.hanno.hpengine.graphics.constants.BufferTarget
import de.hanno.hpengine.graphics.constants.glValue
import org.apache.logging.log4j.LogManager
import org.lwjgl.BufferUtils
import org.lwjgl.opengl.ARBBufferStorage.GL_MAP_COHERENT_BIT
import org.lwjgl.opengl.ARBBufferStorage.GL_MAP_PERSISTENT_BIT
import org.lwjgl.opengl.GL15.*
import org.lwjgl.opengl.GL30
import org.lwjgl.opengl.GL30.GL_MAP_WRITE_BIT
import org.lwjgl.opengl.GL44
import org.lwjgl.opengl.GL44.GL_BUFFER_IMMUTABLE_STORAGE
import org.lwjgl.opengl.GL45.glCopyNamedBufferSubData
import java.nio.ByteBuffer
import kotlin.math.max

private val logger = LogManager.getLogger(PersistentMappedBuffer::class.java)

class PersistentMappedBuffer(
    private val graphicsApi: GraphicsApi,
    override var target: BufferTarget,
    capacityInBytes: SizeInBytes = SizeInBytes(1024)
) : GpuBuffer {
    private var bufferDefinition: BufferDefinition = createBuffer(capacityInBytes)

    override val id: Int get() = bufferDefinition.id
    override val buffer: ByteBuffer get() = bufferDefinition.buffer

    @Synchronized // TODO: remove this
    override fun ensureCapacityInBytes(requestedCapacity: SizeInBytes) {
        val capacityInBytes = SizeInBytes(max(requestedCapacity.value.toInt(), 10))

        val needsResize = SizeInBytes(buffer.capacity()) < capacityInBytes
        if (needsResize) {
            logger.debug("Resizing buffer $id to $capacityInBytes")
            bufferDefinition = createBuffer(capacityInBytes)
            logger.debug("Resized buffer $id")
        }
    }

    override fun put(src: ByteBuffer) {
        buffer.rewind()
        buffer.put(src)
        buffer.rewind()
    }
    private fun BufferDefinition.copyTo(newBuffer: BufferDefinition) {
        graphicsApi.onGpu {
            glCopyNamedBufferSubData(this.id, newBuffer.id, 0, 0, this.buffer.capacity().toLong())
        }
    }

    private fun createBuffer(capacityInBytes: SizeInBytes): BufferDefinition = graphicsApi.onGpu {
        when(val currentDefinition = bufferDefinition) {
            null -> {
                val id = glGenBuffers()
                glBindBuffer(target.glValue, id)
                GL44.glBufferStorage(target.glValue, capacityInBytes.value, flags)
                val byteBuffer = GL30.glMapBufferRange(
                    target.glValue,
                    0, capacityInBytes.value, flags,
                    BufferUtils.createByteBuffer(max(10, capacityInBytes.value.toInt()))
                )!!
                val isImmutable = glGetBufferParameteri(target.glValue, GL_BUFFER_IMMUTABLE_STORAGE) == GL_TRUE
                BufferDefinition(id, byteBuffer, isImmutable)
            }
            else -> {
                logger.debug("Old buffer ${currentDefinition.id}")
                if(currentDefinition.isImmutable) {
                    val id = glGenBuffers()
                    glBindBuffer(target.glValue, id)
                    GL44.glBufferStorage(target.glValue, capacityInBytes.value, flags)
                    val byteBuffer = GL30.glMapBufferRange(
                        target.glValue,
                        0, capacityInBytes.value, flags,
                        BufferUtils.createByteBuffer(max(10, capacityInBytes.value.toInt()))
                    )!!
                    val isImmutable = glGetBufferParameteri(target.glValue, GL_BUFFER_IMMUTABLE_STORAGE) == GL_TRUE

                    currentDefinition.buffer.copyTo(byteBuffer)
                    glDeleteBuffers(currentDefinition.id)

                    BufferDefinition(id, byteBuffer, isImmutable)
                } else {
                    val byteBuffer = GL30.glMapBufferRange(
                        target.glValue,
                        0, capacityInBytes.value, flags,
                        currentDefinition.buffer
                    )!!
                    currentDefinition.buffer.copyTo(byteBuffer)
                    BufferDefinition(id, byteBuffer, currentDefinition.isImmutable)
                }
            }
        }
    }
    override fun bind() = graphicsApi.onGpu {
        glBindBuffer(target.glValue, id)
    }

    override fun unbind() = graphicsApi.onGpu { glBindBuffer(target.glValue, 0) }
    override fun delete() = graphicsApi.onGpu {
        glDeleteBuffers(id)
    }

    private companion object {
        const val flags = GL_MAP_WRITE_BIT or GL_MAP_PERSISTENT_BIT or GL_MAP_COHERENT_BIT
    }
}
